flytekit/common/tasks/sagemaker/distribution.py

```diff
@@ -0,0 +1,99 @@
+import json as _json
+import os as _os
+
+import retry as _retry
+
+from flytekit.common.tasks import sdk_runnable as _sdk_runnable
+from flytekit.engines import common as _common_engine
+
+SM_RESOURCE_CONFIG_FILE = "/opt/ml/input/config/resourceconfig.json"
+SM_ENV_VAR_CURRENT_HOST = "SM_CURRENT_HOST"
+SM_ENV_VAR_HOSTS = "SM_HOSTS"
+SM_ENV_VAR_NETWORK_INTERFACE_NAME = "SM_NETWORK_INTERFACE_NAME"
+
+
+@_retry.retry(exceptions=KeyError, delay=1, tries=10, backoff=1)
+def get_sagemaker_distributed_training_context_from_env() -> dict:
+    distributed_training_context = {}
+    if (
+        not _os.environ.get(SM_ENV_VAR_CURRENT_HOST)
+        or not _os.environ.get(SM_ENV_VAR_HOSTS)
+        or _os.environ.get(SM_ENV_VAR_NETWORK_INTERFACE_NAME)
+    ):
+        raise KeyError
+
+    distributed_training_context[DistributedTrainingContextKey.CURRENT_HOST] = _os.environ.get(SM_ENV_VAR_CURRENT_HOST)
+    distributed_training_context[DistributedTrainingContextKey.HOSTS] = _json.loads(_os.environ.get(SM_ENV_VAR_HOSTS))
+    distributed_training_context[DistributedTrainingContextKey.NETWORK_INTERFACE_NAME] = _os.environ.get(
+        SM_ENV_VAR_NETWORK_INTERFACE_NAME
+    )
+
+    return distributed_training_context
+
+
+@_retry.retry(exceptions=FileNotFoundError, delay=1, tries=10, backoff=1)
```

Comment:
why the retries?

Reply:
The files we try to read here is mounted by SageMaker so it is not guaranteed to be immediately available